### PR TITLE
Refactor alternative engine API

### PR DESCRIPTION
**Pull request by frankfliu**
1. make NDArrayAdapter base class
2. Fallback to alternative engine when needed
3. Allow NDArray auto convert between engines
4. Prepare for ops across different engine

Change-Id: Ia1499169c432d7a76e2df550a266b61240a55609

## Description ##

Brief description of what this PR is about

- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

**Review by lanking520 — APPROVED** [review]

**Comment by lanking520** [comment]
This change may introduce some latency increment from hidden conversion. Saying if I import two engines at the same time in my packages:

- MXNet
- PaddlePaddle

If I would like to use PaddlePaddle alone without touching other packages, it may by default create NDArray in MXnet and convert to Paddle.

**Comment by frankfliu** [comment]
> This change may introduce some latency increment from hidden conversion. Saying if I import two engines at the same time in my packages:
> 
> * MXNet
> * PaddlePaddle
> 
> If I would like to use PaddlePaddle alone without touching other packages, it may by default create NDArray in MXnet and convert to Paddle.

With this change, it will create PaddlePaddle NDArray, not alternative engine, alternative engine is kick in only when the operation that the engine is not supported.

File: api/src/main/java/ai/djl/ndarray/NDManager.java [hunk 1]
```diff
@@ -161,6 +161,14 @@ static NDManager from(NDResource resource) {
      */
     ByteBuffer allocateDirect(int capacity);
 
+    /**
+     * Converts an external engine's {@link NDArray} to current engine.
```
Comment (zachgk) [inline api/src/main/java/ai/djl/ndarray/NDManager.java:165]: Does this convert or create a new one? Both description and adopt seem to imply changing.

Maybe rename to `matching(array)`.

File: api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java [hunk 2]
```diff
@@ -27,908 +28,1029 @@
...
+    public NDArray sequenceMask(NDArray sequenceLength) {
+        return sequenceMask(sequenceLength, 0);
     }
 
     /** {@inheritDoc} */
     @Override
-    default NDArray zerosLike() {
-        throw new UnsupportedOperationException(UNSUPPORTED_MSG);
+    public NDArray zerosLike() {
+        return getManager().zeros(getShape(), getDataType(), getDevice());
```
Comment (zachgk) [inline api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:205]: Some of these are fully generic and don't rely on anything adapter specific. We may want to move the definitions to NDArray instead. This could be in another PR if you prefer.

File: api/src/main/java/ai/djl/ndarray/NDManager.java (hunk 1 above)
Comment (frankfliu) [reply api/src/main/java/ai/djl/ndarray/NDManager.java:165]: How about `from(NDArray)` ?

File: api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java (hunk 2 above)
Comment (frankfliu) [reply api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:205]: I didn't move to NDArray is because I'm not sure if we should remove native implementation for MXNet and PyTorch

File: api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java [hunk 3]
```diff
@@ -27,908 +28,1029 @@
...
+    public NDArray onesLike() {
+        return getManager().ones(getShape(), getDataType(), getDevice());
     }
 
     /** {@inheritDoc} */
     @Override
-    default boolean contentEquals(Number number) {
-        throw new UnsupportedOperationException(UNSUPPORTED_MSG);
+    public boolean contentEquals(Number number) {
+        return number.equals(toArray()[0]);
```
[the remark this answers is not in the document]
Comment (frankfliu) [reply api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:217]: Will fix it in next PR.

File: api/src/main/java/ai/djl/ndarray/NDManager.java (hunk 1 above)
Comment (zachgk) [reply api/src/main/java/ai/djl/ndarray/NDManager.java:165]: Yeah, `from` is better. Go with that

File: api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java (hunk 2 above)
Comment (zachgk) [reply api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:205]: We don't need to remove the native implementations. The ones in NDArray are just defaults and can be overridden for better performance

Comment (frankfliu) [reply api/src/main/java/ai/djl/ndarray/NDArrayAdapter.java:205]: I don't think there is performance gain, the only thing it might saved is getShape() and. getDataType() call.
